File: rust/herb-linter/src/rules/parser_no_errors.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};

use herb::ParseResult;
use herb_config::Severity;

pub struct ParserNoErrorsRule;

impl Rule for ParserNoErrorsRule {
  fn name(&self) -> &'static str {
    "parser-no-errors"
  }

  fn default_severity(&self) -> Severity {
    Severity::Error
  }
}

impl ParserRule for ParserNoErrorsRule {
  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    result
      .recursive_errors()
      .into_iter()
      .map(|error| UnboundOffense::new(self.name(), format!("{} (`{}`)", error.message(), error.error_type()), error.location().clone()))
      .collect()
  }
}

File: javascript/packages/linter/src/rules/parser-no-errors.ts
import { ParserRule } from "../types.js"

import type { LintOffense, FullRuleConfig } from "../types.js"
import type { ParseResult, HerbError } from "@herb-tools/core"

export class ParserNoErrorsRule extends ParserRule {
  static ruleName = "parser-no-errors"
  static introducedIn = this.version("0.5.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  check(result: ParseResult): LintOffense[] {
    return result.recursiveErrors().map(error =>
      this.herbErrorToLintOffense(error)
    )
  }

  protected herbErrorToLintOffense(error: HerbError): LintOffense {
    return {
      ...super.herbErrorToLintOffense(error),
      message: `${error.message} (\`${error.type}\`)`,
    }
  }
}
